### DISCUSSION TITLE
Wavelength-dependent phase function

### DISCUSSION THREAD

**Discussion by DoreenMulder** [mitsuba-renderer/mitsuba3#735] [Q&A]
Hello,

I am trying to implement a wavelength-dependent phase function, but I'm running into some issues I'm not entirely sure how to solve.

I'm using Mitsuba in its `llvm_ad_spectral_polarized` mode, since I would like to make use of automatic differentiation in a later stage of my program. The llvm backend assigns multiple wavelengths to a single ray to speed up the rendering process. 

I pass a ray to my phase function. Since the phase function I aim to use is wavelength dependent, I get a different `(wo, pdf)` for each wavelength. But these wavelengths are all associated to the same ray, which only has one direction. I'm wondering if anyone has advice for me as to what would be best to do in this case? 

While I could use a scalar variant to operate on a single wavelength for each ray, I need the llvm backend for its automatic differentiation later. Is it possible to tell Mitsuba to only use one wavelength per ray without having to change over to a scalar variant? 

Many thanks!

Doreen

**Comment by njroussel** [comment]
Hi @DoreenMulder 

As you're using a pretty complex variant (`llvm_ad_spectral_polarized`), I'm going to assume you're somewhat familiar with how to compile/build Mitsuba. In the `mitsuba.conf` file, spectral variants will typically define their corlor representation as `Spectrum<Float, 4>`. The `4` here defines the number of sampled wavelengths per ray. You should be able to modify it directly. Quite honestly, as far as I know, I don't know how well this feature is supported, but it's worth trying.

There has also been some work to properly support spectral phase functions, check out this PR: https://github.com/mitsuba-renderer/mitsuba3/pull/748. This is a much more robust approach.
